# Working with the Auto MPG Data Set

In this post we will look into the Auto MPG data set and clean it so that it is ready for further use.

This data set shows the mpg of a group of car models produced in the 1970s and the 1980s along with some characteristic information associated with each model. More information about the data set can be found here.

# 1. Download and Load Data File
Download the auto-mpg.data data file and save it to the local directory where you would run the code on this page.

In [70]:
import pandas as pd

In [189]:
df = pd.read_csv('http://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data',  delim_whitespace=True, names=["mpg", "cylinders", "displacement", "horsepower", "weight", "acceleration", "model_year", "origin", "car_name"])
print(df.shape)
df.head()

(398, 9)


mpg  cylinders  displacement horsepower  weight  acceleration  model_year  \
0  18.0          8         307.0      130.0  3504.0          12.0          70   
1  15.0          8         350.0      165.0  3693.0          11.5          70   
2  18.0          8         318.0      150.0  3436.0          11.0          70   
3  16.0          8         304.0      150.0  3433.0          12.0          70   
4  17.0          8         302.0      140.0  3449.0          10.5          70   

   origin                   car_name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

The data has 9 columns (features) and 398 entries. Of the 9 columns, 8 are numerical. The labeling on the origin column is 1 for domestic, 2 for Europe and 3 for Asia. This data set is typically used to learn and predict the mpg column using the remaining columns.

# 2. Cleaning the Data

Upon close inspection we find that there 5 missing values in the horsepower column and that they have been entered as "?"s:

In [191]:
df[df.horsepower=='?']#to found missing values

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?  2046.0          19.0   
126  21.0          6         200.0          ?  2875.0          17.0   
330  40.9          4          85.0          ?  1835.0          17.3   
336  23.6          4         140.0          ?  2905.0          14.3   
354  34.5          4         100.0          ?  2320.0          15.8   
374  23.0          4         151.0          ?  3035.0          20.5   

     model_year  origin              car_name  
32           71       1            ford pinto  
126          74       1         ford maverick  
330          80       2  renault lecar deluxe  
336          80       1    ford mustang cobra  
354          81       2           renault 18i  
374          82       1        amc concord dl

In [192]:
df.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model_year        int64
origin            int64
car_name         object
dtype: object

We have another problem. The horsepower column appears as an object. We need to convert it to float or int.

Replacing the missing values with the mean / median / mode is a crude way of treating missing values. Depending on the context, like if the variation is low or if the variable has low leverage over the response, such a rough approximation is acceptable and could possibly give satisfactory results. To fix the missing value problem, we can fill in those locations with the median (if there are outliers in the data), the mean value of the horsepower data (if there aren't impact of outliers).

To be able to do that we first assign NaN to the missing data locations, change the data type of the horsepower column to float and then assign each missing location the mean value of the column.

In [193]:
#df.loc[horsepower_missing_ind, 'horsepower'] = float('nan')
#df.horsepower = df.horsepower.apply(pd.to_numeric)
#df.loc[horsepower_missing_ind, 'horsepower'] = int( df.horsepower.mean() )

If the missing values in low percentile comparing population we can delete it. In this project, we will delete the missing values because the percentile of missing value is very low comparing rest of data.

We will delete the missing rows and transform horsepower object to numeric

In [194]:
df = df[df.horsepower!="?"]
df[['horsepower']]=df[['horsepower']].apply(pd.to_numeric)

In [195]:
df.dtypes

mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
model_year        int64
origin            int64
car_name         object
dtype: object

Now, things are in order and the statistics should come out fine. We can use the dataframe.describe() function to get a quick idea of the statistics of the data.

In [196]:
df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  392.000000  392.000000    392.000000  392.000000   392.000000   
mean    23.445918    5.471939    194.411990  104.469388  2977.584184   
std      7.805007    1.705783    104.644004   38.491160   849.402560   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.000000    4.000000    105.000000   75.000000  2225.250000   
50%     22.750000    4.000000    151.000000   93.500000  2803.500000   
75%     29.000000    8.000000    275.750000  126.000000  3614.750000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year      origin  
count    392.000000  392.000000  392.000000  
mean      15.541327   75.979592    1.576531  
std        2.758864    3.683737    0.805518  
min        8.000000   70.000000    1.000000  
25%       13.775000   73.000000    1.000000  
50%       15.500000   76.000000    1.000000  
75%       17.025000   79.000000    2.000000  
max       24.800000   82.000000    3.000000

# 3. Saving the Data as a .CSV File

The data will be saved to a .csv due to its versatility.

In [ ]:
#df.to_csv('auto-mpg.csv', index=False)